tarofreighter/chain_porter.go

```diff
@@ -1089,6 +1112,54 @@ func (p *ChainPorter) stateStep(currentPkg sendPackage) (*sendPackage, error) {
 	}
 }
 
+// RegisterSubscriber adds a new subscriber to the set of subscribers that will
+// be notified of any new events that are broadcast.
+func (p *ChainPorter) RegisterSubscriber(
+	receiver *chanutils.EventReceiver[Event],
+	deliverExisting bool, deliverFrom bool) error {
+
+	p.subscriberMtx.Lock()
+	defer p.subscriberMtx.Unlock()
+
+	if p.subscribers == nil {
+		p.subscribers = make(map[uint64]*chanutils.EventReceiver[Event])
+	}
+	p.subscribers[receiver.ID()] = receiver
+
+	return nil
+}
+
+// RemoveSubscriber removes a subscriber from the set of subscribers that will
+// be notified of any new events that are broadcast.
+func (p *ChainPorter) RemoveSubscriber(
+	subscriber *chanutils.EventReceiver[Event]) error {
+
+	p.subscriberMtx.Lock()
+	defer p.subscriberMtx.Unlock()
+
+	_, ok := p.subscribers[subscriber.ID()]
+	if !ok {
+		return fmt.Errorf("subscriber with ID %d not found",
+			subscriber.ID())
+	}
+
+	subscriber.Stop()
+	delete(p.subscribers, subscriber.ID())
+
+	return nil
+}
+
+// publishSubscriberEvent publishes an event to all subscribers.
+func (p *ChainPorter) publishSubscriberEvent(event Event) {
+	for _, sub := range p.subscribers {
```

Comment:
Need to hold the lock over the subscribers here as well to avoid it being changed while we send out updates.